flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStoreTest.java

```diff
@@ -0,0 +1,306 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.jobmanager;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.runtime.checkpoint.TestingRetrievableStateStorageHelper;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.state.RetrievableStateHandle;
+import org.apache.flink.runtime.statehandle.StateHandleStore;
+import org.apache.flink.runtime.statehandle.TestingStateHandleStore;
+import org.apache.flink.util.AbstractID;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.runtime.statehandle.StateHandleStore.NON_EXIST_RESOURCE_VERSION;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link DefaultJobGraphStore} with {@link TestingJobGraphStoreWatcher}, {@link TestingStateHandleStore}
+ * and {@link TestingJobGraphListener}.
+ */
+public class DefaultJobGraphStoreTest extends TestLogger {
+
+	private final JobGraph testingJobGraph = new JobGraph();
+	private final long timeout = 3 * 1000;
+
+	private TestingStateHandleStore.Builder<JobGraph> builder;
+	private TestingRetrievableStateStorageHelper<JobGraph> jobGraphStorageHelper;
+	private TestingJobGraphStoreWatcher testingJobGraphStoreWatcher = new TestingJobGraphStoreWatcher();
+	private TestingJobGraphListener testingJobGraphListener = new TestingJobGraphListener();
+
+	@Before
+	public void setup() {
+		builder = TestingStateHandleStore.builder();
+		jobGraphStorageHelper = new TestingRetrievableStateStorageHelper<>();
+	}
+
+	@After
+	public void teardown() {
+		if (testingJobGraphStoreWatcher != null) {
+			testingJobGraphStoreWatcher.stop();
+		}
+	}
+
+	@Test
+	public void testRecoverJobGraph() throws Exception {
+		final RetrievableStateHandle<JobGraph> stateHandle = jobGraphStorageHelper.store(testingJobGraph);
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setGetFunction(ignore -> stateHandle)
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+
+		final JobGraph recoveredJobGraph = jobGraphStore.recoverJobGraph(testingJobGraph.getJobID());
+		assertThat(recoveredJobGraph, is(notNullValue()));
```

Comment:
```
assertThat(recoveredJobGraph, notNullValue());
```

Reply:
Hmm. I used to believe that `assertThat(recoveredJobGraph, is(notNullValue()))` is more like a sentence. So I keep the is if we could.